### PR TITLE
Lucene indexation : convention on null field values

### LINKED ISSUE #2778: Lucene indexation : convention on null field values
By playing around with the Lucene Queries I've noticed that when we are not providing a value for a text field ; it will not index it even if we explicitly configure it to be "stored". The issue I had with this is that if I want to filter my content items by using a ElasticSearch Wildcard Query then it will return nothing.

https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-wildcard-query.html

So, as stated in the documentation of ElasticSearch here : 
https://www.elastic.co/guide/en/elasticsearch/guide/current/_dealing_with_null_values.html
We should then normally index a "NULL" value that is decided by convention.

Right now I've modified simply the TextFieldIndexHandler and the ContentPickerFieldIndexHandler so that it index a "NULL" string.

```C#
using System.Threading.Tasks;
using OrchardCore.ContentFields.Fields;
using OrchardCore.Indexing;

namespace OrchardCore.ContentFields.Indexing
{
    public class TextFieldIndexHandler : ContentFieldIndexHandler<TextField>
    {
        public override Task BuildIndexAsync(TextField field, BuildFieldIndexContext context)
        {
            var options = context.Settings.ToOptions();

            foreach (var key in context.Keys)
            {
                context.DocumentIndex.Set(key, field.Text ?? "NULL", options);
            }

            return Task.CompletedTask;
        }
    }
}
```

The question is : **should we set a value convention here or is there known alternatives for this through the usage of ElasticSearch Queries?**

## Repro

Create a field and enable indexing + store option. Create an item of that content type without providing any value for the TestField.

image

Lucene Query without "NULL" value indexed (what we currently have in dev branch) : 

image

On the picture underneat ; we can see that I have content items if I'm not filtering by using a wildcard.

image

Though if I want to filter by saying that I want all Acme content items that have a field with ANY value : 

image

I get no results because as you can see nothing is indexed for that field. Though, normally this Lucene Query should return all content items still.

---

Here is the result if I index a default "NULL" string value on Acme.TestField.

image

### PR DESCRIPTION
**Pull request by Skrypt**
WIP

Fixes #2778

### CODE REVIEW

**Comment by Skrypt** [comment]
Need to test with String.Empty instead of NULL.

File: src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/ContentPickerFieldIndexHandler.cs [hunk 1]
```diff
@@ -10,11 +10,20 @@ public override Task BuildIndexAsync(ContentPickerField field, BuildFieldIndexCo
...
+                    {
+                        context.DocumentIndex.Set(key, contentItemId, options);
+                    }
+                }
+            }
+            else {
                 foreach (var key in context.Keys)
                 {
-                    context.DocumentIndex.Set(key, contentItemId, options);
+                    context.DocumentIndex.Set(key, "NULL", options);
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/ContentPickerFieldIndexHandler.cs]: I think the field should not be set if there is a null value, for any type of field. Would that work?

Comment (Skrypt) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/ContentPickerFieldIndexHandler.cs]: I understand what you mean, I need to try different things here. Somehow I required a NULL value to be indexed to be able to find content items that had a NULL value set. I need to re-run those Queries and evaluate.

Comment (sebastienros) [reply src/OrchardCore.Modules/OrchardCore.ContentFields/Indexing/ContentPickerFieldIndexHandler.cs]: And with that we could use queries like this:

https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-exists-query.html

**Comment by Skrypt** [comment]
Closing as we should use Exists queries.

**Comment by Skrypt** [comment]
Ok, we can't with Lucene search for a Field that has not been indexed. So, the workaround explained on the ElasticSearch documentation and also on StackOverflow is to set a convention on NULL values and to index them or to add a value on the Field itself to explain that this field is not empty like explained there : 

https://stackoverflow.com/questions/2732290/lucene-net-find-documents-that-do-not-contain-a-specified-field

Once we will have decided which way to go, which should be quite equal in size term for the Index itself we could after that create a ExistsQuery that would return only items with Fields that have that "NULL" value.

**Comment by Skrypt** [comment]
Here is the result of a wildcard query when indexing a String.Empty as a NULL convention : 

image

Here we see that I have 2 content items.

image

Here if I'm trying to filter by any values I get only 1 result which is the one that has an actual value indexed.

File: src/OrchardCore.Modules/OrchardCore.Taxonomies/Indexing/TaxonomyFieldIndexHandler.cs [hunk 2]
```diff
@@ -46,9 +46,15 @@ public override async Task BuildIndexAsync(TaxonomyField field, BuildFieldIndexC
 
             foreach (var key in context.Keys)
             {
-                foreach (var contentItem in inheritedContentItems)
+                if (inheritedContentItems.Count > 0)
                 {
-                    context.DocumentIndex.Set(key + ".Inherited", contentItem.ContentItemId, options);
+                    foreach (var contentItem in inheritedContentItems)
+                    {
+                        context.DocumentIndex.Set(key + ".Inherited", contentItem.ContentItemId ?? "NULL", options);
```
Comment (sebastienros) [inline src/OrchardCore.Modules/OrchardCore.Taxonomies/Indexing/TaxonomyFieldIndexHandler.cs]: This can never be null

**Comment by sebastienros** [comment]
Why not set the `"NULL"` value in Lucene class that creates the STRING|TEXT entry directly?

**Comment by Skrypt** [comment]
You remembered me why I kept this as a draft.

**Comment by Skrypt** [comment]
Moved to https://github.com/OrchardCMS/OrchardCore/pull/4062

**Pull request closed without merging**